src/ZkFold/Base/Protocol/ARK/Plonk.hs

```diff
@@ -3,227 +3,157 @@
+import           Prelude                                    hiding (Num (..), div, drop, length, replicate, sum, take,
+                                                             (!!), (/), (^))
+import qualified Prelude                                    as P hiding (length)
+import           Test.QuickCheck                            (Arbitrary (..))
 
 import           ZkFold.Base.Algebra.Basic.Class
-import           ZkFold.Base.Algebra.Basic.Field                     (Zp)
 import           ZkFold.Base.Algebra.Basic.Number
-import           ZkFold.Base.Algebra.Basic.Permutations              (fromPermutation)
-import           ZkFold.Base.Algebra.EllipticCurve.BLS12_381         (BLS12_381_G1, BLS12_381_G2, BLS12_381_Scalar)
-import           ZkFold.Base.Algebra.EllipticCurve.Class             (EllipticCurve (..), Pairing (..), Point)
-import           ZkFold.Base.Algebra.Polynomials.Univariate          hiding (qr)
-import           ZkFold.Base.Data.Vector                             (Vector (..), fromVector)
-import           ZkFold.Base.Protocol.ARK.Plonk.Internal             (getParams)
-import           ZkFold.Base.Protocol.ARK.Plonk.Relation             (PlonkRelation (..), toPlonkRelation)
-import           ZkFold.Base.Protocol.Commitment.KZG                 (com)
+import           ZkFold.Base.Algebra.Basic.Permutations     (fromPermutation)
+import           ZkFold.Base.Algebra.EllipticCurve.Class    (EllipticCurve (..), Pairing (..), Point)
+import           ZkFold.Base.Algebra.Polynomials.Univariate hiding (qr)
+import           ZkFold.Base.Data.Vector                    (Vector (..), fromVector)
+import           ZkFold.Base.Protocol.ARK.Plonk.Internal
+import           ZkFold.Base.Protocol.ARK.Plonk.Relation    (PlonkRelation (..), toPlonkRelation)
+import           ZkFold.Base.Protocol.Commitment.KZG        (com)
 import           ZkFold.Base.Protocol.NonInteractiveProof
-import           ZkFold.Prelude                                      (length, take, (!))
-import           ZkFold.Symbolic.Compiler.ArithmeticCircuit.Internal (ArithmeticCircuit (..), inputVariables)
-
--- TODO (Issue #25): make this module generic in the elliptic curve with pairing
-
-type F = Zp BLS12_381_Scalar
-type G1 = Point BLS12_381_G1
-type G2 = Point BLS12_381_G2
+import           ZkFold.Prelude                             (length, (!))
+import           ZkFold.Symbolic.Compiler                   (Arithmetic, ArithmeticCircuit, inputVariables)
 
 {-
     NOTE: we need to parametrize the type of transcripts because we use BuiltinByteString on-chain and ByteString off-chain.
     Additionally, we don't want this library to depend on Cardano libraries.
 -}
 
--- TODO: disallow `d` that are not powers of 2
-data Plonk (d :: Natural) (n :: Natural) t = Plonk F F F (Vector n Natural) (ArithmeticCircuit 1 F) F
-    deriving (Show)
+data Plonk (n :: Natural) (l :: Natural) curve1 curve2 transcript = Plonk {
+        omega :: ScalarField curve1,
```

Comment:
hmmm `ScalarField curve1` is `BLS12_381_G1` not a `F`
`type PlonkBS n = Plonk PlonkPolyLengthBS n BLS12_381_G1 BLS12_381_G2 ByteString`

Reply:
No, `ScalerField BLS12_381_G1` is `Zp BLS12_381_Scalar` as before. See `BLS12_381.hs`.